menu: ringback/early audio handling for parallel calls

cspiel1 opened the pull request:
This PR introduces a logic for handling ringback tones and early media for multiple outgoing calls (parallel call feature).

- call,menu: no early audio if multiple outgoing calls
- menu: delayed play for ringback and ring tone

Follow up PR:
- These changes: https://github.com/baresip/baresip/pull/2368 . Check if SIP UPDATE is allowed directly in `call_modify()` and return err if not allowed.

alfredh commented on src/call.c:
```diff
@@ -2322,12 +2337,12 @@ static void sipsess_progr_handler(const struct sip_msg *msg, void *arg)
 	else {
-		call_stream_stop(call);
 		call_event_handler(call, CALL_EVENT_RINGING, call->peer_uri);
+		call_stream_stop(call);
```
Not sure about this.

The handler may deref the call object.
When returning it will operate on a invalid call object.

cspiel1 replied:
Right, I have to analyze old commits. We had this problem once. There might be a general problem if an event handler is allowed to deref the call. Especially, if there are multiple event handlers.

cspiel1 replied:
Would it be a good idea to:
```c
	mem_ref(call);
	if (media) {
		call_event_handler(call, CALL_EVENT_PROGRESS, call->peer_uri);
		call_stream_start(call, false);
		mem_deref(call);
	}
	else {
		call_event_handler(call, CALL_EVENT_RINGING, call->peer_uri);
		call_stream_stop(call);
		mem_deref(call);
	}
```

cspiel1 replied:
This https://github.com/baresip/baresip/pull/2403/commits/16a28ce26b77a4d39d1dd5caf66f145996f8cf0a ensures that the `call_stream_start()` does not operate on a destroyed call object.

alfredh commented on modules/menu/menu.c:
```diff
@@ -23,6 +23,7 @@ static struct menu menu;
 
 enum {
 	MIN_RINGTIME = 1000,
+	TONE_DELAY   =   20,
```
perhaps add a comment about which unit is used

alfredh commented on modules/menu/menu.c:
```diff
@@ -117,6 +118,35 @@ static char *errorcode_key_aufile(uint16_t scode)
+		call_set_audio_ldir(call, ardir & SDP_SENDONLY);
+		/* TODO: 100rel used and allowed */
+/*                if (call_refresh_allowed(call))*/
+/*                        call_modify(call);*/
```
perhaps uncomment the code or remove commented out code before commit

alfredh commented on modules/menu/menu.c:
```diff
@@ -304,21 +334,38 @@ static void play_ringback(const struct call *call)
+
+	ardir = sdp_media_rdir(stream_sdpmedia(audio_strm(call_audio(call))));
+	ring = (ardir & SDP_RECVONLY) ? false : true;
+	if (ring && !menu.ringback &&
```
adding empty newline before `if` makes the code more reable

alfredh commented on modules/menu/menu.c:
```diff
@@ -304,21 +334,38 @@ static void play_ringback(const struct call *call)
+	enum sdp_dir ardir;
+
+	ardir = sdp_media_rdir(stream_sdpmedia(audio_strm(call_audio(call))));
+	ring = (ardir & SDP_RECVONLY) ? false : true;
```
`ardir` and `ring` can be declared and set at the same time

also:

`ring = !(ardir & SDP_RECVONLY)`

alfredh commented on modules/menu/menu.h:
```diff
@@ -32,6 +32,7 @@ struct menu{
 	int32_t adelay;               /**< Outgoing auto answer delay     */
 	char *ansval;                 /**< Call-Info/Alert-Info value     */
 	struct odict *ovaufile;       /**< Override aufile dictionary     */
+	struct tmr tmr_play;          /**< Tones play timer               */
```
did you forget to call `tmr_cancel` ?

cspiel1 replied:
Indeed. I add it to `menu_stop_play()`.

alfredh commented:
what is the next step here ? I see that the PR is marked as "Draft".

cspiel1 commented:
I added a commit: https://github.com/baresip/baresip/pull/2403/commits/4de3d4de237c68dd33706d342acbb5c2457dc780
And rebased on current main.

I'll repreat the tests for parallel outgoing calls. Then will mark as "ready for review".

cspiel1 commented:
In a next PR: 
Check if SIP UPDATE is allowed directly in `call_modify()` and return err if not allowed. Or a level deeper, in `sipsess_modify()`? Or even in `sipsess_update()`?

@alfredh, @sreimers Let me know what you think about this!
